0000-settlement-engines/0000-settlement-engines.md

```diff
@@ -0,0 +1,404 @@
+
+[Account for an incoming settlement](#account-for-incoming-settlements). The settlement engine sends this request as it receives incoming settlements in in order to ensure [settlement symmetry](#settlement-symmetry). If the accounting system responds that it only credited a partial amount, the settlement engine [tracks the uncredited amounts](#track-uncredited-incoming-settlements).
+
+#### Request
+
+```http
+POST /accounts/:id/settlements HTTP/1.1
+Accept: application/json
+Content-Type: application/json
+Idempotency-Key: <key>
+```
+
+> **[`Quantity`](#quantity-json-type)** to be credited to the account as an incoming settlement
+
+#### Response
+
+```http
+HTTP/1.1 201 Created
+Content-Type: application/json
+```
+
+> **[`Quantity`](#quantity-json-type)** credited to the account as an incoming settlement, which MUST  be less than or equal to the quantity in the original request
+
+### Send outgoing message
+
+Send the message to the given peer's settlement engine and return its response. The connector handles [proxying the message](#exchanging-messages) through the peer's connector.
+
+#### Request
+
+```http
+POST /accounts/:id/messages HTTP/1.1
+Accept: application/octet-stream
+Content-Type: application/octet-stream
+```
+
+> _&lt;raw bytes of message&gt;_
+
+#### Response
+
+```http
+HTTP/1.1 201 OK
+Content-Type: application/octet-stream
+```
+
+> _&lt;raw bytes of response&gt;_
+
+## Idempotence
+
+Idempotent requests ensure each side effect only happens once, even though the same request may be called multiple times. For example, if a request to perform a settlement fails due to a network connection error, [idempotence](https://en.wikipedia.org/wiki/Idempotence) prevents a client from accidentally triggering multiple settlements when it retries the request.
+
+### Performing idempotent requests
+
+Requests to settle or requests to credit incoming settlements MUST include an idempotency key, or globally unique string, within an `Idempotency-Key: <key>` header. To avoid collisions, this key MUST be derived from a cryptographically secure source of randomness.
+
+### Retry behavior
+
+This retry behavior ensures the client and server are eventually consistent and never perform any unsafe balance rollbacks that could result in lost funds or double payments.
+
+If the client receives no response, an HTTP 5xx error, or an HTTP 409 Conflict error, the client MUST retry the request with the same idempotency key. If the response is a client error, such as another HTTP 4xx error, the client MUST rollback the balance update and NOT perform any subsequent retries.
+
+To prevent overwhelming the server, the client SHOULD exponentially backoff after each failed retry attempt and add random "jitter" to vary the retry interval. The maximum retry interval MUST be no greater than 1 hour. Clients also MUST retry indefinitely until they have received an acknowledgement from the server that the request was processed or failed.
```

Comment:
```suggestion
To prevent overwhelming the server, the client SHOULD exponentially backoff after each failed retry attempt and add random "jitter" to vary the retry interval. The maximum retry interval MUST be no greater than 1 hour. Clients also MUST retry indefinitely until they have received an acknowledgment from the server that the request was processed or failed.
```